src/prop/sat_proof_manager.h

```diff
@@ -0,0 +1,559 @@
+ *
+ * In the above example this happens as: step S3
+ *
+ *    (or l1 l2)   ~l1    ~l2
+ *  --------------------------- CHAIN_RES
+ *             false
+ *
+ * is added for a conflict clause (or l1 l2). Then we attempt to explain ~l1 and
+ * ~l2. The former has no explanation (i.e., it's an input), while the latter
+ * has explanation (or l3 ~l2). This leads to step S4
+ *
+ *  (or l3 ~l2)     ~l3
+ *  -------------------- CHAIN_RES
+ *        ~l2
+ *
+ * being added to the LazyCDProofChain. In explaining ~l3 the step S5
+ *
+ *  (or l1 ~l3)      ~l1
+ *  --------------------- CHAIN_RES
+ *         ~l3
+ *
+ * is added. Since ~l1 has no explanation, the process halts. Note that at this
+ * point the step S6 has not been added to the LazyCDProofChain. This is because
+ * to explain ~l5 we need to look at the literal premises in the proofs of
+ * learned clauses.
+ *
+ * The last thing done then in case (3), after the resolution on the conflict
+ * clause and an initial recursive explanation of propagated literals, is to
+ * connect the refutation proof and then recursively its propagated literal
+ * leaves, repeating until a fix point.
+ *
+ * In the above example the first connection yields
+ *
+ *                           (or l1 ~l3 l5)   ~l5
+ *                           ---------------------- CHAIN_RES
+ *                                     (or l1 ~l3)              ~l1
+ *                                     ----------------------------- CHAIN_RES
+ *                      (or l3 ~l2)     ~l3
+ *                      -------------------- CHAIN_RES
+ *  (or l1 l2)   ~l1    ~l2
+ * --------------------------- CHAIN_RES
+ *                     false
+ *
+ * whose literal leaves are ~l1 and ~l5. The former has no explanation, but the
+ * latter is explained by (or l1 ~l5), thus leading to step S6
+ *
+ *  (or l1 ~l5)      ~l1
+ *  --------------------- CHAIN_RES
+ *         ~l5
+ *
+ * then the final connection
+ *
+ *                                             (or l1 ~l5)  ~l1
+ *                                             ---------------- CHAIN_RES
+ *                            (or l1 ~l3 l5)   ~l5
+ *                            ---------------------- CHAIN_RES
+ *                                      (or l1 ~l3)              ~l1
+ *                                      ----------------------------- CHAIN_RES
+ *                       (or l3 ~l2)     ~l3
+ *                       -------------------- CHAIN_RES
+ *   (or l1 l2)   ~l1    ~l2
+ *  --------------------------- CHAIN_RES
+ *                      false
+ *
+ * which has no more explainable literals as leaves.
+ *
+ */
+class SatProofManager
+{
+ public:
+  SatProofManager(Minisat::Solver* solver,
+                  TheoryProxy* proxy,
+                  context::UserContext* userContext,
+                  ProofNodeManager* pnm);
+
+  /** Marks the start of a resolution chain.
+   *
+   * The given clause, at the node level, is registered in d_resLinks with a
```

Comment:
I assume this is connected to endResChain?  Can you say the expectation about the timeline of when these are called?  E.g. it is expected that there is a call of endResChain that follows startResChain after X.